core/data_array.cpp

```diff
@@ -223,58 +211,88 @@ Variable sparse_dense_op_impl(const VariableConstView &sparseCoord_,
                  }});
 }
 
+namespace {
+bool is_self_append(const DataArrayConstView &a, const DataArrayConstView &b) {
+  return &a.underlying() == &b.underlying();
+}
+} // namespace
+
 DataArray &DataArray::operator+=(const DataArrayConstView &other) {
   expect::coordsAreSuperset(*this, other);
   union_or_in_place(masks(), other.masks());
-  data() += other.data();
+  if (hasData() && other.hasData()) {
+    data() += other.data();
+  } else {
+    if (is_self_append(*this, other))
+      event::append(unaligned(), copy(other.unaligned()));
+    else
+      event::append(unaligned(), other.unaligned());
+  }
   return *this;
 }
 
 DataArray &DataArray::operator-=(const DataArrayConstView &other) {
   expect::coordsAreSuperset(*this, other);
   union_or_in_place(masks(), other.masks());
-  data() -= other.data();
+  if (hasData() && other.hasData()) {
+    data() -= other.data();
+  } else {
+    const DataArray tmp =
+        is_self_append(*this, other) ? copy(other.unaligned()) : DataArray{};
+    auto other_unaligned = tmp ? DataArrayConstView(tmp) : other.unaligned();
+    unaligned().data() *= -1.0;
```

Comment:
I expected to see the sign flip on `other_unaligned`. Since you flip the sign again at the end it has the same effect. Though leave this `DataArray` in a bad state if `event::append` throws.

Reply:
I wanted to avoid making a copy, this is why it is written as it is.

I agree about the problem with exception safety. Unfortunately we by now have many cases like this, and we will need to do a cleanup at some point, trying to reduce these issues.

In this case in particular I'll suggest leaving it for now, since we may want to write it differently for performance reasons soon.